starknet-providers/src/jsonrpc/models/mod.rs

```diff
@@ -591,6 +681,196 @@ pub struct BlockHashAndNumber {
+    DeployAccount(BroadcastedDeployAccountTransaction),
+}
+
+/// Possible values for version are 0 or 1. This may effect the possible parameters for StarkNet
+/// RPCs and their resulting transaction hashes. (see
+/// https://docs.starknet.io/documentation/architecture_and_concepts/Blocks/transactions/ for more
+/// info)
+#[serde_as]
+#[derive(Debug, Clone)]
+pub enum TransactionVersion {
+    V0,
+    V1,
+}
+
+/// (`BROADCASTED_INVOKE_TXN`) Mempool representation of an invoke transaction
+#[derive(Debug, Clone)]
+pub struct BroadcastedInvokeTransaction {
+    /// The maximal fee that can be charged for including the transaction
+    pub max_fee: FieldElement,
+    pub signature: Vec<FieldElement>,
+    pub nonce: FieldElement,
+    pub versioned_properties: InvokeTransactionProperties,
+}
+
+/// (`BROADCASTED_DECLARE_TXN`) Mempool representation of a declare transaction.
+#[serde_as]
+#[derive(Debug, Clone, Serialize)]
+pub struct BroadcastedDeclareTransaction {
+    pub version: TransactionVersion,
+    /// The maximal fee that can be charged for including the transaction
+    #[serde_as(as = "UfeHex")]
+    pub max_fee: FieldElement,
+    #[serde_as(as = "Vec<UfeHex>")]
+    pub signature: Vec<FieldElement>,
+    #[serde_as(as = "UfeHex")]
+    pub nonce: FieldElement,
+    /// The class to be declared
+    pub contract_class: ContractClass,
+    /// The address of the account contract sending the declaration transaction
+    #[serde_as(as = "UfeHex")]
+    pub sender_address: FieldElement,
+}
+
+/// (`BROADCASTED_DEPLOY_TXN`) Mempool representation of a deploy transaction. The structure of a
+/// deploy transaction. Note that this transaction type is deprecated and will no longer be
+/// supported in future versions
+#[serde_as]
+#[derive(Debug, Clone, Serialize)]
+pub struct BroadcastedDeployTransaction {
+    /// Version of the transaction scheme
+    pub version: TransactionVersion,
+    /// The class of the contract that will be deployed
+    pub contract_class: ContractClass,
+    /// The salt for the address of the deployed contract
+    #[serde_as(as = "UfeHex")]
+    pub contract_address_salt: FieldElement,
+    /// The parameters passed to the constructor
+    #[serde_as(as = "Vec<UfeHex>")]
+    pub constructor_calldata: Vec<FieldElement>,
+}
+
+/// (`BROADCASTED_DEPLOY_ACCOUNT_TXN`) Mempool representation of a deploy account transaction
+#[serde_as]
+#[derive(Debug, Clone, Serialize)]
+pub struct BroadcastedDeployAccountTransaction {
+    /// The maximal fee that can be charged for including the transaction
+    #[serde_as(as = "UfeHex")]
+    pub max_fee: FieldElement,
+    /// Version of the transaction scheme
+    pub version: TransactionVersion,
+    #[serde_as(as = "Vec<UfeHex>")]
+    pub signature: Vec<FieldElement>,
+    #[serde_as(as = "UfeHex")]
+    pub nonce: FieldElement,
+    /// The salt for the address of the deployed contract
+    #[serde_as(as = "UfeHex")]
+    pub contract_address_salt: FieldElement,
```

Comment:
I don't like how this allows people to construct invalid transaction requests: you can specify `version` to be `V0` without encountering a compile-time error. However there's actually no v0 for `DeployAccount` transaction type.

Reply:
Understand we should do this. I'm a bit confused about which transaction versions the types should allow to be constructed here. Not sure the [docs.starknet.io](https://docs.starknet.io/documentation/architecture_and_concepts/Blocks/transactions/#declare_transaction) and jsonrpc spec are giving me a full picture. Should this just target v0.10.3 and should constructing transaction version 0 for declare/invoke/deploy transactions be allowed? And also now that adding deploy transactions are deprecated from v0.10.3, should the add_deploy_transaction method be removed?

Reply:
I general we don't support deprecated features since they will be removed from the protocol completely (without needing to support old block data) eventually with the regenesis. However, I want to make an exception here for JSON RPC as we're specifically targeting a certain specification. Making our code closer to the specs helps with maintenance.